File: services/auth-service/database_config.go
package main

import (
	"fmt"
	"net"
	"net/url"
	"os"
)

// databaseDSN preserves DATABASE_URL for local development and also supports
// split settings so Kubernetes can inject a Vault-backed password without
// assembling a connection string containing credentials.
func databaseDSN() (string, error) {
	if databaseURL := os.Getenv("DATABASE_URL"); databaseURL != "" {
		return databaseURL, nil
	}

	required := map[string]string{
		"DB_HOST":     os.Getenv("DB_HOST"),
		"DB_NAME":     os.Getenv("DB_NAME"),
		"DB_USER":     os.Getenv("DB_USER"),
		"DB_PASSWORD": os.Getenv("DB_PASSWORD"),
	}
	for name, value := range required {
		if value == "" {
			return "", fmt.Errorf("%s deve ser definida quando DATABASE_URL não for usada", name)
		}
	}

	port := os.Getenv("DB_PORT")
	if port == "" {
		port = "5432"
	}
	sslMode := os.Getenv("DB_SSLMODE")
	if sslMode == "" {
		sslMode = "require"
	}

	dsn := &url.URL{
		Scheme: "postgresql",
		User:   url.UserPassword(required["DB_USER"], required["DB_PASSWORD"]),
		Host:   net.JoinHostPort(required["DB_HOST"], port),
		Path:   required["DB_NAME"],
	}
	query := dsn.Query()
	query.Set("sslmode", sslMode)
	dsn.RawQuery = query.Encode()

	return dsn.String(), nil
}

File: services/flag-service/database_config.py
import os


def build_database_config(environ=None):
    """Return psycopg2 pool arguments without embedding a secret in a URL."""
    env = os.environ if environ is None else environ

    database_url = env.get("DATABASE_URL")
    if database_url:
        return {"dsn": database_url}

    required = ("DB_HOST", "DB_NAME", "DB_USER", "DB_PASSWORD")
    missing = [name for name in required if not env.get(name)]
    if missing:
        raise ValueError(
            f"{', '.join(missing)} deve(m) ser definida(s) quando DATABASE_URL não for usada"
        )

    try:
        port = int(env.get("DB_PORT", "5432"))
    except ValueError as error:
        raise ValueError("DB_PORT deve ser um número inteiro") from error

    return {
        "host": env["DB_HOST"],
        "port": port,
        "dbname": env["DB_NAME"],
        "user": env["DB_USER"],
        "password": env["DB_PASSWORD"],
        "sslmode": env.get("DB_SSLMODE", "require"),
    }
